### ISSUE TITLE
As Stream creator I should be able to specify upsert behavior for mongodb sink

### ISSUE THREAD

**Issue by mbhuie** [spring-cloud-stream-app-starters/mongodb#4]
As stream creator I should be able to specify upsert behavior for mongodb sink using a natural key.

Acceptance:

Given an inbound messages
```
{
  "orderId" : "OR0001", 
  "type" : "kiosk",
  "item" : "tea"
},
{
  "orderId" : "OR0002", 
  "type" : "desk",
  "item" : "coffee"
},
{
  "orderId" : "OR0001", 
  "type" : "desk",
  "item" : "tea"
}
```
And key column = "orderId"
And upsert = true
When mongo sink recieves these messages in above sequence
Then the collection in mongodb should look like below
```
{
  "orderId" : "OR0001", 
  "type" : "desk",
  "item" : "tea"
},
{
  "orderId" : "OR0002", 
  "type" : "desk",
  "item" : "coffee"
}
```

**Comment by artembilan** [comment]
No sure in your concern: that is exactly how MongoDB sink works.
It is based on the `MongoDbStoringMessageHandler` which, in turn, performs `this.mongoTemplate.save()` operation:
```
	/**
	 * Save the object to the specified collection. This will perform an insert if the object is not already present, that
	 * is an 'upsert'.
	...
	 */
	void save(Object objectToSave, String collectionName);
```

The downstream logic in the `saveDocument()` is like:
```
if (!dbDoc.containsKey(ID_FIELD)) {
	if (writeConcernToUse == null) {
		collection.insertOne(dbDoc);
	} else {
		collection.withWriteConcern(writeConcernToUse).insertOne(dbDoc);
	}
} else if (writeConcernToUse == null) {
	collection.replaceOne(Filters.eq(ID_FIELD, dbDoc.get(ID_FIELD)), dbDoc, new UpdateOptions().upsert(true));
} else {
	collection.withWriteConcern(writeConcernToUse).replaceOne(Filters.eq(ID_FIELD, dbDoc.get(ID_FIELD)), dbDoc,
			new UpdateOptions().upsert(true));
}
```

So, maybe you just need to ensure that your document has an `_id` (`MongoTemplate.ID_FIELD`) property ?

**Comment by artembilan** [comment]
Consider this `id` rules BTW: https://docs.spring.io/spring-data/mongodb/docs/1.10.7.RELEASE/reference/html/#mongo-template.id-handling:

> A property or field annotated with `@Id` (`org.springframework.data.annotation.Id`) will be mapped to the `_id` field.

**Comment by mbhuie** [comment]
This forces me to use externally managed ids and create a function to map natural keys in incoming message to mongodb id.

Based on your example it seems like its not possible.

Btw - I was able to create a custom sink for this. It will be good to have it as starter app;

```
MongoCollection<Document> collection = database .getCollection(properties.getCollection());
Document filter = new Document();
for(String key : properties.getKey()) {
  filter.append(key, document.get(key));
}    	
UpdateResult updateResult = collection.replaceOne(filter, document, new UpdateOptions().upsert(properties.isUpsert()));
```

Note that it does not use mongoTemplate as Spring Data MongoDB is yet to catchup revised mongodb Document API.

**Comment by artembilan** [comment]
After some debating in my hear I came to the conclusion that I won't to see your improvement as a contribution. Only the concern that I wonder how we are going to keep existing functionality. 
Or, of course, we can come up with the same `@ConditionalOnProperty` to do regular `this.mongoTemplate.save(payload, collectionName);` or this `upsert` variant...

**Comment by malinovskiy-alex** [comment]
Hey @artembilan 
So are you going to implement this enhancement?

**Comment by artembilan** [comment]
Well, as I said before: the `MongoTemplate` supports an `upsert`, but only if your domain object comes with an `id` property:
```
collection.replaceOne(mapped.getIdFilter(), dbDoc, new ReplaceOptions().upsert(true));
```

So, we have everything on board. Only what you need to do is to follow domain specific requirements to achieve the goal.

**Comment by malinovskiy-alex** [comment]
Yeah, but I don't want to add my domain object in the sink app and I think that solution which @mbhuie  has suggested is more generic...

**Comment by artembilan** [comment]
OK. I would like to see a sample how you are going to use this assuming that there is an `upsert` support.

Thanks

**Comment by malinovskiy-alex** [comment]
So in order to implement it, I need to update class MongoDbStoringMessageHandler. Am I allowed to suggest changes in this class?

**Comment by artembilan** [comment]
Not sure what are you talking about.

I asked you about the sample how you are going to use this feature in the SCDF stream definition. What do you have to make this MongoDB Sink to make an upsert instead of insert?
We are not talking about any changes yet.
First of all we need to figure out what is a use-case and how it is going to be so generic that any other solutions may gain of it.
